python-packages/aws-event-stream/aws_event_stream/_private/deserializers.py

```diff
@@ -0,0 +1,112 @@
+            case "exception":
+                member_name = expect_type(str, headers[":exception-type"])
+                consumer(schema.members[member_name], message_deserializer)
+            case "error":
+                raise UnexpectedEventError(
+                    expect_type(str, headers[":error-code"]),
+                    expect_type(str, headers[":error-message"]),
+                )
+            case _:
+                raise EventError(f"Unknown event structure: {event}")
+
+
+class EventMessageDeserializer(SpecificShapeDeserializer):
+    def __init__(
+        self, headers: HEADERS_DICT, payload_deserializer: ShapeDeserializer
+    ) -> None:
+        self._headers = headers
+        self._payload_deserializer = payload_deserializer
+
+    def read_struct(
+        self,
+        schema: Schema,
+        consumer: Callable[[Schema, ShapeDeserializer], None],
+    ) -> None:
+        headers_deserizlizer = EventHeaderDeserializer(self._headers)
+        for key in self._headers.keys():
+            member_schema = schema.members.get(key)
+            if member_schema is not None and EVENT_HEADER_TRAIT in member_schema.traits:
+                consumer(member_schema, headers_deserizlizer)
+
+        if (payload_member := self._get_payload_member(schema)) is not None:
+            consumer(payload_member, self._payload_deserializer)
+        else:
+            self._payload_deserializer.read_struct(schema, consumer)
+
+    def _get_payload_member(self, schema: "Schema") -> "Schema | None":
+        for member in schema.members.values():
+            if EVENT_PAYLOAD_TRAIT in member.traits:
+                return member
+        return None
+
+
+class EventHeaderDeserializer(SpecificShapeDeserializer):
+    def __init__(self, headers: HEADERS_DICT) -> None:
+        self._headers = headers
+
+    def read_boolean(self, schema: "Schema") -> bool:
+        return expect_type(bool, self._headers[schema.expect_member_name()])
+
+    def read_blob(self, schema: "Schema") -> bytes:
+        return expect_type(bytes, self._headers[schema.expect_member_name()])
+
+    def read_byte(self, schema: "Schema") -> int:
+        return expect_type(int, self._headers[schema.expect_member_name()])
+
+    def read_short(self, schema: "Schema") -> int:
+        return expect_type(int, self._headers[schema.expect_member_name()])
+
+    def read_integer(self, schema: "Schema") -> int:
+        return expect_type(int, self._headers[schema.expect_member_name()])
+
+    def read_long(self, schema: "Schema") -> int:
+        return expect_type(int, self._headers[schema.expect_member_name()])
+
+    def read_string(self, schema: "Schema") -> str:
+        return expect_type(str, self._headers[schema.expect_member_name()])
+
+    def read_timestamp(self, schema: "Schema") -> datetime.datetime:
+        # TODO: do we support timestamp format here? One would assume not since the
+        # format has a specific timestamp type.
```

Comment:
This would be specifying the timestamp format as any one of our three supported formats: epoch seconds, http date, or RFC 3339 (`date-time`). We of course use this in HTTP headers, json bodies, etc. But those formats don't have a native timestamp type. `application/vnd.amazon.eventstream` does have a native timestamp type (which ends up being epoch *milli*seconds plus a type signifier).

Personally, I would prefer we don't allow customization of the format here. I'm not sure if we can get away with that though.

Reply:
I'm in favor of starting with that. If we see significant feedback, adding this later is much easier than removing it.